cardinal/server/events.go

```diff
@@ -1,32 +1,159 @@
 package server
 
 import (
-	"log"
 	"net/http"
+	"time"
+
+	"github.com/go-openapi/runtime/middleware"
 
 	"github.com/gorilla/websocket"
+	"github.com/rs/zerolog/log"
 )
 
+func CreateEventHub() *EventHub {
+	res := EventHub{
+		websocket_connections: map[*websocket.Conn]bool{},
+		Broadcast:             make(chan []byte),
+		Register:              make(chan *websocket.Conn),
+		Unregister:            make(chan *websocket.Conn),
+		Shutdown:              make(chan bool),
+	}
+	return &res
+}
+
+type Event struct {
+	message string
+}
+
+type EventHub struct {
+	websocket_connections map[*websocket.Conn]bool
+	Broadcast             chan []byte
+	Unregister            chan *websocket.Conn
+	Register              chan *websocket.Conn
+	Shutdown              chan bool
+}
+
+func (h *EventHub) run() {
+Loop:
+	for {
+		select {
+		case conn := <-h.Register:
+			h.websocket_connections[conn] = true
+		case conn := <-h.Unregister:
+			if _, ok := h.websocket_connections[conn]; ok {
+				delete(h.websocket_connections, conn)
+				err := conn.Close()
+				if err != nil {
+					log.Logger.Error().Err(err)
+				}
+			}
+		case message := <-h.Broadcast:
+			for conn := range h.websocket_connections {
+
+				err := conn.SetWriteDeadline(time.Now().Add(3 * time.Second))
+				if err != nil {
+					go func() {
+						h.Unregister <- conn
+					}()
+					log.Logger.Error().Err(err)
+					break Loop
+				}
+				err = conn.WriteMessage(websocket.TextMessage, message)
+				if err != nil {
+					go func() {
+						h.Unregister <- conn
+					}()
+					log.Logger.Error().Err(err)
+				}
+
+			}
+		case <-h.Shutdown:
+			go func() {
+				for conn, _ := range h.websocket_connections {
+					h.Unregister <- conn
```

Comment:
This connection will never be unregistered. break Loop is called directly after spinning up this goroutine, so this select statement will never be used again.

Reply:
hmm yeah good point. Will fix.